### PR TITLE
feat: model declaration with decorators

### PR DESCRIPTION
**Pull request by cyjake**
- https://www.yuque.com/leoric/blog/sgifi3 (zh-cn)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,7 +11,8 @@
   ],
   "scripts": {
     "jsdoc": "rm -rf docs/api && jsdoc -c .jsdoc.json -d docs/api -t node_modules/@cara/minami",
-    "pretest": "./test/prepare.sh",
+    "prepack": "tsc",
+    "pretest": "tsc && ./test/prepare.sh",
```
Comment (cyjake) [inline package.json:15]: make sure tsc is executed before pack/publish or test

File: src/bone.js [hunk 2]
```diff
@@ -1116,7 +1126,7 @@ class Bone {
   static hasOne(name, opts) {
     opts = Object.assign({
       className: capitalize(name),
-      foreignKey: this.table + 'Id',
+      foreignKey: camelCase(`${this.name}Id`),
```
[the remark this answers is not in the document]
Comment (cyjake) [reply src/bone.js]: updated

File: src/decorators.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import DataType from './data_types';
+import { ASSOCIATE_METADATA_MAP } from './constants';
+import 'reflect-metadata';
+
+type DataTypes<T> = {
+  [Property in keyof T as Exclude<Property, "toSqlString">]: T[Property];
+}
+
+interface ColumnOption {
```
[the remark this answers is not in the document]
Comment (cyjake) [reply src/decorators.ts:10]: will leave this to a separate decorator, something like `@Validate`, in another pr

File: src/bone.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (cyjake) [reply src/bone.js:1129]: define.underscored was never meant to address name conventions in javascript side, it was used to configure the field name convention at database side.

**Review by JimmyDaddy — APPROVED** [review]

**Pull request merged**